### ISSUE TITLE
Query engine library for current platform "rhel-openssl-1.1.x" could not be found.

### ISSUE THREAD

**Issue by simonNozaki** [prisma/prisma#13518]
### Bug description

# Outline
Our node.js apps encountered the error on this title. Full message is below:
```
Query engine library for current platform "rhel-openssl-1.1.x" could not be found.
You incorrectly pinned it to rhel-openssl-1.1.x
```

## About our app
Our app uses prisma client on fetching a record from PostgreSQL(by AWS RDS) 1:1 by HTTP GET request. GET request endpoint is like this:
```
https://hash.execute-api.ap-northeast-1.amazonaws.com/stage/path?id=<id>
```

# Situation
The error happened when a Client worker executing regularly calls the REST API using prisma client over and over again. We can find that the database load (connections, memory usage, cpu utilizations and so on) is not much hard.

### How to reproduce

Sorry but, we can not find the way to reproduce this error(so that we are hard to solve this problem...).

## schema.prisma
```
generator client {
  provider = "prisma-client-js"
  binaryTargets = ["native", "rhel-openssl-1.0.x"]
  previewFeatures = ["interactiveTransactions"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model table_name {
    ...
}
```

## prisma client code
The mapper implements with prisma client is like this:
```ts
async select(id: string): Promise<schema[]> {
    const prisma = new PrismaClient()
    let condition = {
        where: {
            id: id,
            requester_id: requesterId
        }
    }
    
    const result = await prisma.schema.findMany(condition)
    prisma.$disconnect()
    return result
}
```

## package.json
```json
  },
  "dependencies": {
    ...
    "@prisma/client": "^3.6.0",
    ...
  },
  "devDependencies": {
    ...
    "prisma": "^3.6.0",
    ...
  },
}
```

### Environment & setup

We are deploying node.js app using prisma on AWS Lambda.
- Lambda: node.js: 14.x
- AWS Lambda

### Prisma Version

```
$ ./node_modules/.bin/prisma -v
prisma                  : 3.8.0
@prisma/client          : 3.8.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 34df67547cf5598f5a6cd3eb45f14ee70c3fb86f (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 34df67547cf5598f5a6cd3eb45f14ee70c3fb86f (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 34df67547cf5598f5a6cd3eb45f14ee70c3fb86f (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 34df67547cf5598f5a6cd3eb45f14ee70c3fb86f (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : 34df67547cf5598f5a6cd3eb45f14ee70c3fb86f
Studio                  : 0.452.0
```

**Comment by jkomyno** [comment]
Hi @simonNozaki, would you please be able to try again using a more modern version of `prisma`, like `prisma@3.14.0`?

**Comment by pantharshit00** [comment]
If you can share your project structure here, that would be awesome.

**Comment by simonNozaki** [comment]
@jkomyno @pantharshit00 
Thank you for quick replyings!

> Hi @simonNozaki, would you please be able to try again using a more modern version of `prisma`, like `prisma@3.14.0`?
Yes, I'll try it.

> If you can share your project structure here, that would be awesome.
Okay, directories of our app is below. I can not understand of your needs, so please tell me if you want more information of this.
```
README.md
aws
buildspec.yml
docker-compose.yml
jest.config.js
local
node_modules
package-lock.json
package.json
prisma
  └　migrations
    └　20220114065029_init
    └　20220201001432_rename_column
    └ migration_lock.toml
  └schema.prisma
serverless.yml
src
test
tsconfig.json
tsconfig.paths.json
webpack.config.js
```

**Comment by pantharshit00** [comment]
Looks like you are using the serverless framework. Make sure you are also using the latest version of their tooling.

**Comment by simonNozaki** [comment]
> Looks like you are using the serverless framework. Make sure you are also using the latest version of their tooling.

I see, I'll do that.

**Comment by simonNozaki** [comment]
> Looks like you are using the serverless framework. Make sure you are also using the latest version of their tooling.

I did this, but... we encountered severe errors like this...Of course, we googled about this error, but could not build, too...

```
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
```

we are hard to solve this

**Comment by pantharshit00** [comment]
I will recommend following this guide: https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-aws-lambda

**Comment by simonNozaki** [comment]
I did not know that there are official documents combining AWS Lambda! Thanks, I'll share that with my team and refine settings around prisma.

**Comment by janpio** [comment]
Does this happen to _all_ requests sending this query or only to some? 
Or asked another way: Do some requests for exactly the same data return successfully while others do not?

**Comment by simonNozaki** [comment]
> Does this happen to _all_ requests sending this query or only to some? Or asked another way: Do some requests for exactly the same data return successfully while others do not?

It seems that it does not depend on data, but yesterday could find that this error occur when clients requests for our app many times continuously (like looping 1000times).

For example, seeing logs, around 850 requests were successful randomly.

**Comment by simonNozaki** [comment]
> That would indicate that there are some Lambdas where the file can be found, and some where it can not be - as some function executions succeed, and some don't. As you essentially deploy exactly the same code to all of them, that does not really make a lot of sense 🤔

Thank you for replying. We finally give up using prisma client on our app. Instead of prisma/client, we adopted `sequelize` and that is now  stably working.

Thanks for all contributors to your support. I could learn about prisma/client.

**Comment by gpadvorac** [comment]
Hopefully this helps someone.  I had the same problem.  in our case we had a recursive AWS serverless lambda function that processed a 1mb file from S3 and imported its data into mysql.  the function would import only one file and then invoke another instance of itself to process the next file.  We have our reasons for this type of process.  The error we got was:

```
Invalid `prisma.service.findFirst()` invocation in
/var/task/index.js:34610:46

34607 // services/common/createSessionForService.ts
34608 var MOD_FUNCTION = "services.common.createSessionForService.ts";
34609 async function createSessionForService(prisma, serviceName) {
→ 34610 const serviceResult = await prisma.service.findFirst(
Query engine library for current platform "rhel-openssl-1.1.x" could not be found.
You incorrectly pinned it to rhel-openssl-1.1.x
```
The lambda had 128 mb of memory and got this error after processing 3 files.

we increased the memory to 512mb and we got this error after processing 60 files.

We increased the memory to 2048mb and processed 1200 files (loops) (calling await prisma.service.findFirst 1200 times) and all 1200 processed with no error.

So my conclusion is this error is related to memory or lack of...

**Comment by gpadvorac** [comment]
At the end of the function code it has invokes its self again.  Recursive.  Same function, same environment.  It pulls a file from S3 and imports the data into a database.  As I increased the size of the files or the number of files I would get this error, and sometimes a different error, but the other error (I forget what) seemed to be memory related according to my searches.  As I increased the memory these error went away.

The reason for invoking it recursively is if one lambda call processes the entire batch, it could time out as we have no control of how big the batches will be.

With 1.5GB memory we processed 1200 files @860kb each without fail.
Also 110 files @ 25MB each without fail

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by janpio** [comment]
Ah, in theory each next triggered function _could_ of course be the _same_ function environment - as this is effectively a sequential row of function executions. If the function leaks some memory, the next run would obviously already start with some memory in use and then hit a limit. (Still weird that this would manifest as the error you reported...)

Do you do `new PrismaClient` inside or outside of your function "handler"? Does the function connect to the database once for each function invocation or only on cold start?

**Comment by gpadvorac** [comment]
Yes I do a new PrismaClient inside the function and it connects to the database and does all work inside the function.  The function is independent this way and I think it would release resources when done.  And by doing it this way only one instance of the function is running at a time.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by janpio** [comment]
Are you closing the connection by calling `$disconnect()` at the end of each function? Otherwise each function "instance" will create a new, additional connection at the end of each function execution.

**Comment by GermaVinsmoke** [comment]
Do we have any document for AWS Lambda layers? I added the dependencies of the project as a layer and used the arn of that layer in the main serverless project. The application is working but the routes which are having prisma functions are not working.

> Error: \nInvalid `prisma.user.findMany()` invocation:\n\n\nQuery engine library for current platform \"rhel-openssl-1.0.x\" could not be found.\nYou incorrectly pinned it to rhel-openssl-1.0.x\n\nThis probably happens, because you built Prisma Client on a different platform.\n(Prisma Client looked in \"/opt/nodejs/node16/node_modules/@prisma/client/runtime/libquery_engine-rhel-openssl-1.0.x.so.node\")

**Comment by janpio** [comment]
Please open a new issue and provide the information necessary for us to understand what exactly you are doing. A closed issue is not the right place to debug and have this conversation. Thanks.
